docs/src/rfcs/002-shader-stage-interpolants.md

```diff
@@ -0,0 +1,188 @@
+
+For pixel shader outputs in modern GLSL, instead of return a value from a function one writes to a global `out` type. And similarly, shader inputs also come from globals.
+
+### Decorators
+
+Similar to DirectX, Vulkan has the following decorations that can be applied to interpolants.
+
+| Vulkan | DirectX equivalent
+| ---- | ---- |
+| - | linear |
+| Flat| nointerpolation
+| NoPerspective| noperspective
+| Centroid | centroid
+| Sample| sample
+
+Vulkan has some additional explanations about these decorations as well:
+
+> A variable decorated with `Flat` will not be interpolated. Instead, it will have the same value for every fragment within a triangle. This value will come from a single provoking vertex. A variable decorated with `Flat` can also be decorated with Centroid or Sample, which will mean the same thing as decorating it only as `Flat`.
+
+## Metal shading language
+
+__Mostly todo.__
+
+One thing to note as a different to SPIR-V, GLSL and HLSL is that MSL doesn't seem to require explicit binding semantics or `Location` specifiers. It's specification has a `[[attribute(N)]]` specifier, but it's only used on tesselation control points and vertex shader inputs. It's specifically not used for communication between shader stages.
+
+# SPIR-V Practicalities
+
+Vertex shaders allow `Location` and `Component` decorations on input variable declarations. The `Location` decoration specifies which vertex input attribute is used to read and interpret the data that a variable will consume. The `Component` decoration allows the location to be more finely specified for scalars and vectors, down to the individual components within a location that are consumed. The components within a location are 0, 1, 2, and 3. A variable starting at component N will consume components N, N+1, N+2, …​ up through its size. For single precision types, it is invalid if the sequence of components gets larger than 3.
+
+
+## Proposal
+
+```rust
+// these `attribute`'s would match the input assembler setup on the CPU
+struct VertexInput {
+    #[attribute(0)]
+    position: Vec3,
+    #[attribute(1)]
+    normal: Vec3,
+    #[attribute(2)]
+    color: Vec3,
+}
+
+// naming this is hard, since vertex shader won't really know what shader
+// stages will run after it until the pipeline state object gets created
+// in vulkan. a tesselation shader may run after this for example,
+// and that decision won't be known by our compiler.
+struct VertexOutput {
+    #[position]
+    position: Vec3,
+    normal: Vec3,
+    color: Vec3,
+}
+
+fn vs_main(#[stage_in] input: &VertexInput) -> VertexOutput {
+    VertexOutput {
+        position: world_to_clip.mul(input.position),
+        normal: input.normal,
+        color: input.color,
+    }
+}
+
+struct PixelOutput {
+    // albedo and material id packed into same RGBA8 render target
+    #[render_target(0)]
```

Comment:
Should (all) these attributes be `#[spirv(render_target = 0)]` to be consistent with [the rest of the attributes](https://embarkstudios.github.io/rust-gpu/book/attributes.html)?

Reply:
I thought we had consensus that we shouldn't do that #128 and I was under the impression that most of those `#[spirv]` attributes were either temporary or to get at specific implementation specific things from spirv. Not for higher level features.

Reply:
The conclusion of 128 is that we're keeping `spirv` for now, and perhaps we rename it to `#[shader(render_target = 0)]` in the future. Besides, there's significant technical difficulties in supporting `#[render_target(0)]`, and is the reason `#[spirv(...)]` was invented in the first place.

Reply:
Yeah, I agree even if it's not called `spirv` there should be at least one wrapper attribute around the inner attribute. It would get confusing as a user as to who is defined the attribute.

Reply:
Ah right, I was thinking of these along the same lines as the regular rustc [built-in attributes](https://doc.rust-lang.org/reference/attributes.html#built-in-attributes-index). However, if we feel putting these in their own namespace is the best way forward then I think we should do that. I would suggest putting them in another namespace then the `spriv` namespace though, mostly to keep them separate and avoid potential name clashes, but also to keep the new attribute clean and only add things there after some deliberation (instead of straight up forwarding spirv_headers symbols) and so we can avoid naming things after spirv-isms such as `gl_compute` or the confusion between workgroup and subgroup (two separate but very similar concepts in spirv).

Reply:
> I would suggest putting them in another namespace then the `spriv` namespace though, mostly to keep them separate and avoid potential name clashes

I strongly disagree, the conclusion of #128 is that we're keeping things in `#[spirv(..)]` _for now_. If you believe the current system is unclean and has issues like potential name clashes, then let's address those in another MCP, instead of patching on half-fixes like new namespaces.

> so we can avoid naming things after spirv-isms such as `gl_compute`

I am fully aware of the issues with this and have asked feedback on names, and multiple times the conclusion has been that the current system is fine. If we want to change those decisions, let's open another thread.